src/crl.rs

```diff
@@ -46,6 +46,60 @@ pub trait CertRevocationList: Sealed {
         supported_sig_algs: &[&SignatureAlgorithm],
         issuer_spki: &[u8],
     ) -> Result<(), Error>;
+
+    /// Returns true if the CRL can be considered authoritative for the given certificate.
+    ///
+    /// A CRL is considered authoritative for a certificate when:
+    ///   * The certificate issuer matches the CRL issuer and,
+    ///     * The certificate has no CRL distribution points, and the CRL has no issuing distribution
+    ///       point extension.
+    ///     * Or, the certificate has no CRL distribution points, but the the CRL has an issuing
+    ///       distribution point extension with a scope that includes the certificate.
+    ///     * Or, the certificate has CRL distribution points, and the CRL has an issuing
+    ///       distribution point extension with a scope that includes the certificate, and at least
+    ///       one distribution point full name is a URI type general name that can also be found in
+    ///       the CRL issuing distribution point full name general name sequence.
+    ///
+    /// In all other circumstances the CRL is not considered authoritative.
+    fn authoritative_for(&self, cert: &Cert<'_>) -> bool {
+        // In all cases we require that the authoritative CRL have the same issuer
+        // as the certificate. Recall we do not support indirect CRLs.
+        if self.issuer() != cert.issuer() {
+            return false;
+        }
+
+        match (
+            cert.crl_distribution_points(),
+            self.issuing_distribution_point(),
+        ) {
+            // If the certificate has no CRL distribution points, and the CRL has no issuing distribution point,
+            // then we can consider this CRL authoritative based on the issuer matching.
+            (None, None) => true,
+
+            // If the certificate has one or more CRL distribution points, but this CRL does not
+            // have an issuing distribution point then it can't be authoritative for the cert.
+            (Some(_), None) => false,
+
+            // If the certificate has no CRL distribution points, but this CRL has an issuing
+            // distribution point then we check the CRL IDP scope includes the type of certificate
+            // we have, in addition to matching the issuers.
+            (None, Some(crl_idp)) => {
+                IssuingDistributionPoint::from_der(untrusted::Input::from(crl_idp))
+                    .map(|crl_idp| crl_idp.can_contain(cert))
+                    .unwrap_or(false)
+            }
+
+            // If the certificate has one or more CRL distribution points, and this CRL has an
+            // issuing distribution point, then we decide if the CRL is authoritative by
+            // verifying the CRL IDP scope includes the type of certificate we have, and that
+            // the CRL IDP can be considered authoritative for one of the cert DPs.
+            (Some(cert_dps), Some(crl_idp)) => {
```

Comment:
I think I've implemented what you were looking for here - it helped to restructure things so that checking the CRL scope for a cert and checking the IDP against the cert's DPs was done in one function as suggested in later feedback.

I'm going to resolve this conversation but if I missed the mark please unresolve and I'll take another go at it.